### PR TITLE
[CAST-158] Implement Request Too Slow Detection

### PR DESCRIPTION
**Pull request by justincorsha**
Before you submit your PR, make sure you check each of the following:

1. [x] I have read the [Contributing Guidelines](./CONTRIBUTING.md) and signed the [Contributor License Agreement](./CLA.md).
2. [x] My code is clean and ready-to-PR, including being free of lint errors and dead code
3. [x] I have **tested** my change / code and am confident that it works
4. [x] I have filled in the template below completely and accurately
##### Changelog

- Added Request Too Slow data generation to the `generate-pipeline-data.sh` script
- Created RequestTooSlowCard that populates a table with data from long-running requests
- Created request-too-slow endpoint that finds traffic data which is not a websocket or SSE and whose `elapsedTime` returned from kubeshark is longer than sixty seconds
- Fixed postgres indexes
- Fixed background color for cast ui

#### Medium Severity Fault for Requests Between 1m and 3m
Screenshot 2023-08-25 at 2 16 13 PM

#### High Severity Fault for Requests > 3m
https://github.com/corshatech/cast/assets/109618849/4987acba-cd1c-4db7-b7b6-5cebb79455bd

##### Testing

- Ran skaffold to use `generate-pipeline-data.sh` and checked that longer requests were populated in the table
- Sent requests that were long-running and checked that the correct data populated the table
- Sent requests that had attributes of SSE/websockets and checked that they were not added to the table

##### Unit Tests

- Added snapshot test for new RequestTooSlowCard
- Checked that unit tests passed

### CODE REVIEW

File: ui/components/app/Analysis/RequestTooSlowCard.tsx [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    renderCell(params: GridRenderCellParams<string>) {
+      return params.value && <FormattedDate when={params.value}/>
+    },
+    width: 200,
+  },
+  { field: 'Severity',
+    headerName: 'Severity',
+  },
+  { field: 'Elapsed Time',
+    headerName: 'Elapsed Time (ms)',
```
Comment (ArcticLight) [inline ui/components/app/Analysis/RequestTooSlowCard.tsx:37]: Given that the minimum time for this to happen is in minutes, can this be displayed in units of least seconds?

Please display this value in the table (which may come from the backend in ms) in seconds as in `5010 (ms)` displayed as `5.01 (s)`

File: ui/lib/analysis/request_too_slow.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+      data->'request'->>'absoluteURI' as absolute_uri,
+      occurred_at as timestamp,
+      data->'request'->>'method' as method,
+      data->'request'->>'httpVersion' as version,
+      data->'response'->'content'->>'mimeType' as mimeType,
+      data->'request'->'headers'->>'Connection' as connection_header,
+      data->'request'->'headers'->>'Accept' as accept_header
+      from traffic
+  ) as trafficinner
+where elapsed_time > 60000 and method != 'CONNECT' and coalesce(connection_header, '') != 'Upgrade' and version = 'HTTP/1.1' and mimeType != 'text/event-stream' and not accept_header ~ 'text/event-stream'
```
Comment (ArcticLight) [inline ui/lib/analysis/request_too_slow.ts:35]: Please extract 60000 as a constant and re-embed in the query, to make it clearer what this magic value is for and represents, as in:
```ts
/** Request Too Long maximum threshold, (ms) */
const ELAPSED_TOO_LONG_THRESHOLD = "60000";
// ...
`where elapsed_time > ${ELAPSED_TOO_LONG_THRESHOLD}`
```

File: ui/lib/analysis/request_too_slow.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  
+  // Initially sort by request time descending
+  rows.sort((r1, r2) => r2.elapsed_time - r1.elapsed_time)
+
+  const findings = rows.map((row) => {
+    const finding: RequestTooSlow = {
+      type: 'request-too-slow',
+      name: 'Request Too Slow',
+      detectedAt,
+      severity: row.elapsed_time > 180000 ? 'high' : 'medium',
```
Comment (ArcticLight) [inline ui/lib/analysis/request_too_slow.ts:68]: Extract magic constant `180000` to top of file, same as the threshold value from the SQL query.

File: ui/lib/analysis/request_too_slow.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    id: 'request-too-slow',
+    title: 'Request Too Slow',
+    description:
+      'A request was detected that was in-flight for longer than one minute. ' +
+      'An API can be vulnerable to DoS and resource exhaustion attacks, ' + 
+      'which can use up the allocated resources and potentially cause outages. ' +
+      'Ensure that you have implemented resource limits as well as timeouts for ' +
+      'requests to your APIs.',
+    reportedAt: new Date().toISOString(),
+    severity: maxTime >= 180000 ? 'high' : 'medium',
```
Comment (ArcticLight) [inline ui/lib/analysis/request_too_slow.ts:103]: Use the constant you extracted above, here.

**Review by ArcticLight — CHANGES_REQUESTED** [review]

File: ui/lib/analysis/request_too_slow.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+/** runnerPure is the AnalysisFunction free of external dependencies
+ *
+ * This function allows for dependency injection during unit testing
+ */
+export async function runnerPure(
+  query: () => Promise<Row[]>,
+): Promise<Analysis> {
+  const rows = await query();
+  const detectedAt = new Date().toISOString();
+  const maxTime = rows.length > 0 ? rows.reduce((prev, curr) => curr.elapsed_time > prev.elapsed_time ? curr : prev).elapsed_time : 0;
```
Comment (ArcticLight) [inline ui/lib/analysis/request_too_slow.ts:58]:
```suggestion
  const maxTime = (rows ?? []).reduce((prev, curr) => Math.max(prev, curr), 0)
```

**Review by ArcticLight — CHANGES_REQUESTED** [review]

File: ui/lib/analysis/request_too_slow.ts (hunk 2 above)
Comment (justincorsha) [reply ui/lib/analysis/request_too_slow.ts]: Created `ELAPSED_MEDIUM_SEVERITY_THRESHOLD` variable set to 60 seconds.

File: ui/lib/analysis/request_too_slow.ts (hunk 3 above)
Comment (justincorsha) [reply ui/lib/analysis/request_too_slow.ts]: Created `ELAPSED_HIGH_SEVERITY_THRESHOLD` variable set to 60 seconds.

**Review by JasonCorsha — APPROVED** [review]
LGTM

**Review by ArcticLight — APPROVED** [review]

**Pull request merged**